### PR TITLE
0.30 buffer management

### PR DESCRIPTION
**Pull request by i509VCB**
This adds a way for multiple types of WlBuffer to coexist. The interesting part of this is the `ManagedBuffer` type which statically asserts a buffer is managed by Smithay.

This is done in two commits:
- introducing the `buffer` module
- porting shm to use `buffer` module along with renderer changes.

### CODE REVIEW

File: src/wayland/buffer/mod.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    ///
+    /// At this point the buffer is no longer usable by Smithay.
+    fn buffer_destroyed(&mut self, buffer: &ManagedBuffer);
+}
+
+/// The buffer manager.
+///
+/// This type allows a [`WlBuffer`] to have data associated and managed by Smithay.
+#[derive(Debug)]
+pub struct BufferManager(());
```
Comment (PolyMeilex) [inline src/wayland/buffer/mod.rs:44]: If we want to keep this type, wouldn't this be cleaner?
```suggestion
pub struct BufferManager;
```

File: src/wayland/buffer/mod.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
+//! Buffer management utilities.
+//!
+//! This module provides the [`Buffer`] type to represent a [`WlBuffer`] managed by Smithay and data
```
Comment (vberger) [inline src/wayland/buffer/mod.rs:3]:
```suggestion
//! This module provides the [`ManagedBuffer`] type to represent a [`WlBuffer`] managed by Smithay and data
```

File: src/wayland/buffer/mod.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+use crate::utils::UnmanagedResource;
+
+/// Handler trait for associating data with a [`WlBuffer`].
+///
+/// This trait primarily allows compositors to be told when a buffer is destroyed.
+///
+/// # For buffer abstractions
+///
+/// Buffer abstractions (such as [`shm`](crate::wayland::shm)) should require this trait bound in dispatch
+/// requirements to ensure access to the [`BufferManager`].
```
Comment (vberger) [inline src/wayland/buffer/mod.rs:26]: Is this still relevant?

File: src/wayland/buffer/mod.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+{
+    fn request(
+        self: Arc<Self>,
+        _: &mut Handle<D>,
+        data: &mut D,
+        _: ClientId,
+        msg: Message<ObjectId>,
+    ) -> Option<Arc<dyn ObjectData<D>>> {
+        // WlBuffer has a single request which is a destructor.
+        assert_eq!(msg.opcode, 0);
```
Comment (vberger) [inline src/wayland/buffer/mod.rs:128]: Let's make that a `debug_assert!()`, no need to pollute release code with such checks.

File: src/wayland/shm/handlers.rs [hunk 5]
```diff
@@ -96,22 +100,13 @@ where
  * wl_shm_pool
  */
 
-/// User data of WlShmPool
-#[derive(Debug)]
-pub struct ShmPoolUserData {
-    inner: Arc<Pool>,
-}
```
Comment (vberger) [inline src/wayland/shm/handlers.rs:103]: Why did you move this to `mod.rs`?

File: src/backend/renderer/mod.rs [hunk 6]
```diff
@@ -256,6 +256,7 @@ pub trait ImportMemWl: ImportMem {
     /// with an empty list `&[]`, the renderer is allowed to not update the texture at all.
     fn import_shm_buffer(
         &mut self,
+        dh: &mut DisplayHandle<'_>,
         buffer: &wl_buffer::WlBuffer,
```
Comment (vberger) [inline src/backend/renderer/mod.rs:260]: The fact that this function takes a `wl_buffer` instead of a `ManagedBuffer` makes the whole thing pretty awkward. I think we should find a way to allow `ManagedBuffer` to cover EGL buffers as well.

File: src/wayland/buffer/mod.rs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+        /// Data associated with the managed buffer.
+        data: Arc<BufferData>,
+    },
+
+    /// Buffer not managed by Smithay.
+    ///
+    /// There is no user data that may be taken from the buffer.
+    Unmanaged(WlBuffer),
+}
```
Comment (vberger) [inline src/wayland/buffer/mod.rs:134]: I'm wondering if the inner fields of this enum should be public, or if we should wrap it into a unit struct so that `Buffer` effectively is an opaque type from the point of view of the user.

Comment (i509VCB) [reply src/wayland/buffer/mod.rs:134]: Not sure it's worth exposing the inner fields. Okay the user has an object ID and buffer data. It's not exactly clear what to do with the returned data.

If this were to be public, then I'd need to make more details of BufferData public.

Plus I have added functions to get the data in side these enum variants anyways.

Comment (vberger) [reply src/wayland/buffer/mod.rs:134]: Ok, right, my bad, for some reason I though this was actually the public type. :shrug: All is good then.

**Review by PolyMeilex — APPROVED** [review]
